## Importing the Required Packages

In [1]:
import pandas as pd
import numpy as np
import warnings

## Loading  the Dataset

In [2]:
heart_df_2 = pd.read_csv('breast-cancer.csv',header=None)
heart_df_2

0  1      2      3       4       5        6        7        8   \
0      842302  M  17.99  10.38  122.80  1001.0  0.11840  0.27760  0.30010   
1      842517  M  20.57  17.77  132.90  1326.0  0.08474  0.07864  0.08690   
2    84300903  M  19.69  21.25  130.00  1203.0  0.10960  0.15990  0.19740   
3    84348301  M  11.42  20.38   77.58   386.1  0.14250  0.28390  0.24140   
4    84358402  M  20.29  14.34  135.10  1297.0  0.10030  0.13280  0.19800   
..        ... ..    ...    ...     ...     ...      ...      ...      ...   
564    926424  M  21.56  22.39  142.00  1479.0  0.11100  0.11590  0.24390   
565    926682  M  20.13  28.25  131.20  1261.0  0.09780  0.10340  0.14400   
566    926954  M  16.60  28.08  108.30   858.1  0.08455  0.10230  0.09251   
567    927241  M  20.60  29.33  140.10  1265.0  0.11780  0.27700  0.35140   
568     92751  B   7.76  24.54   47.92   181.0  0.05263  0.04362  0.00000   

          9   ...      22     23      24      25       26       27      28  \
0    0.14710  ...  25.380  17.33  184.60  2019.0  0.16220  0.66560  0.7119   
1    0.07017  ...  24.990  23.41  158.80  1956.0  0.12380  0.18660  0.2416   
2    0.12790  ...  23.570  25.53  152.50  1709.0  0.14440  0.42450  0.4504   
3    0.10520  ...  14.910  26.50   98.87   567.7  0.20980  0.86630  0.6869   
4    0.10430  ...  22.540  16.67  152.20  1575.0  0.13740  0.20500  0.4000   
..       ...  ...     ...    ...     ...     ...      ...      ...     ...   
564  0.13890  ...  25.450  26.40  166.10  2027.0  0.14100  0.21130  0.4107   
565  0.09791  ...  23.690  38.25  155.00  1731.0  0.11660  0.19220  0.3215   
566  0.05302  ...  18.980  34.12  126.70  1124.0  0.11390  0.30940  0.3403   
567  0.15200  ...  25.740  39.42  184.60  1821.0  0.16500  0.86810  0.9387   
568  0.00000  ...   9.456  30.37   59.16   268.6  0.08996  0.06444  0.0000   

         29      30       31  
0    0.2654  0.4601  0.11890  
1    0.1860  0.2750  0.08902  
2    0.2430  0.3613  0.08758  
3    0.2575  0.6638  0.17300  
4    0.1625  0.2364  0.07678  
..      ...     ...      ...  
564  0.2216  0.2060  0.07115  
565  0.1628  0.2572  0.06637  
566  0.1418  0.2218  0.07820  
567  0.2650  0.4087  0.12400  
568  0.0000  0.2871  0.07039  

[569 rows x 32 columns]

Loading the field names from the text file and adding them as column names of the loaded dataframe

In [3]:
heart_df=heart_df_2.copy()
field_names = pd.read_csv('field_names.txt',header=None)
heart_df.columns=field_names[0]
heart_df

0          ID diagnosis  radius_mean  radius_sd_error  radius_worst  \
0      842302         M        17.99            10.38        122.80   
1      842517         M        20.57            17.77        132.90   
2    84300903         M        19.69            21.25        130.00   
3    84348301         M        11.42            20.38         77.58   
4    84358402         M        20.29            14.34        135.10   
..        ...       ...          ...              ...           ...   
564    926424         M        21.56            22.39        142.00   
565    926682         M        20.13            28.25        131.20   
566    926954         M        16.60            28.08        108.30   
567    927241         M        20.60            29.33        140.10   
568     92751         B         7.76            24.54         47.92   

0    texture_mean  texture_sd_error  texture_worst  perimeter_mean  \
0          1001.0           0.11840        0.27760         0.30010   
1          1326.0           0.08474        0.07864         0.08690   
2          1203.0           0.10960        0.15990         0.19740   
3           386.1           0.14250        0.28390         0.24140   
4          1297.0           0.10030        0.13280         0.19800   
..            ...               ...            ...             ...   
564        1479.0           0.11100        0.11590         0.24390   
565        1261.0           0.09780        0.10340         0.14400   
566         858.1           0.08455        0.10230         0.09251   
567        1265.0           0.11780        0.27700         0.35140   
568         181.0           0.05263        0.04362         0.00000   

0    perimeter_sd_error  ...  concavity_worst  concave_points_mean  \
0               0.14710  ...           25.380                17.33   
1               0.07017  ...           24.990                23.41   
2               0.12790  ...           23.570                25.53   
3               0.10520  ...           14.910                26.50   
4               0.10430  ...           22.540                16.67   
..                  ...  ...              ...                  ...   
564             0.13890  ...           25.450                26.40   
565             0.09791  ...           23.690                38.25   
566             0.05302  ...           18.980                34.12   
567             0.15200  ...           25.740                39.42   
568             0.00000  ...            9.456                30.37   

0    concave_points_sd_error  concave_points_worst  symmetry_mean  \
0                     184.60                2019.0        0.16220   
1                     158.80                1956.0        0.12380   
2                     152.50                1709.0        0.14440   
3                      98.87                 567.7        0.20980   
4                     152.20                1575.0        0.13740   
..                       ...                   ...            ...   
564                   166.10                2027.0        0.14100   
565                   155.00                1731.0        0.11660   
566                   126.70                1124.0        0.11390   
567                   184.60                1821.0        0.16500   
568                    59.16                 268.6        0.08996   

0    symmetry_sd_error  symmetry_worst  fractal_dimension_mean  \
0              0.66560          0.7119                  0.2654   
1              0.18660          0.2416                  0.1860   
2              0.42450          0.4504                  0.2430   
3              0.86630          0.6869                  0.2575   
4              0.20500          0.4000                  0.1625   
..                 ...             ...                     ...   
564            0.21130          0.4107                  0.2216   
565            0.19220          0.3215                  0.1628   
566            0.30940          0.3403    

## Data Preprocessing

Checking the data types of the loaded dataset and converting as required. For our ML modelling purpose we will require the diagnosis to be in the type of integer

In [4]:
print(heart_df.dtypes)

0
ID                              int64
diagnosis                      object
radius_mean                   float64
radius_sd_error               float64
radius_worst                  float64
texture_mean                  float64
texture_sd_error              float64
texture_worst                 float64
perimeter_mean                float64
perimeter_sd_error            float64
perimeter_worst               float64
area_mean                     float64
area_sd_error                 float64
area_worst                    float64
smoothness_mean               float64
smoothness_sd_error           float64
smoothness_worst              float64
compactness_mean              float64
compactness_sd_error          float64
compactness_worst             float64
concavity_mean                float64
concavity_sd_error            float64
concavity_worst               float64
concave_points_mean           float64
concave_points_sd_error       float64
concave_points_worst          float64
symmetry_m

The ID column is a unique columns or a name column and is non informative. Thus we will remove this from our data for modelling

In [5]:
del heart_df['ID']

Checking the shape of the loaded dataset to make sure the dataset has been loaded properly.

In [6]:
heart_df.shape

(569, 31)

We will integer encode the target variable (in our case diagnosis). We will replace 'M' as 1 as we will move forward modelling malignant tumors and 'B' as 0. After replacing we can see that there are 357 Benign and 212 Malign diagnosis in the dataframe.

In [7]:
level_mapping= {'M': 1, 'B': 0}
heart_df['diagnosis'] = heart_df['diagnosis'].replace(level_mapping)
heart_df['diagnosis'].value_counts()

0    357
1    212
Name: diagnosis, dtype: int64

Checking for null values in the dataset. Since there are no null values present no action is tobe taken

In [8]:
print(heart_df.isnull().sum())

0
diagnosis                     0
radius_mean                   0
radius_sd_error               0
radius_worst                  0
texture_mean                  0
texture_sd_error              0
texture_worst                 0
perimeter_mean                0
perimeter_sd_error            0
perimeter_worst               0
area_mean                     0
area_sd_error                 0
area_worst                    0
smoothness_mean               0
smoothness_sd_error           0
smoothness_worst              0
compactness_mean              0
compactness_sd_error          0
compactness_worst             0
concavity_mean                0
concavity_sd_error            0
concavity_worst               0
concave_points_mean           0
concave_points_sd_error       0
concave_points_worst          0
symmetry_mean                 0
symmetry_sd_error             0
symmetry_worst                0
fractal_dimension_mean        0
fractal_dimension_sd_error    0
fractal_dimension_worst       0
dtype:

Next step is to remove outliers.Since the data is that of diagnosis of tumor we cannot really say if the bigger values in a column are outliers or not. Thus for this modelling purpose we will go with the same data we have without removing outliers. If the information is given or known we can go forward with removing outliers which are 3-4 standard deviations away by either dropping these rows or imputing the values with a mean or median value. For this dataset we will remove the rows with very far away values only (6 or 7 standard deviations away from the mean value.

In [10]:
from scipy import stats
z_scores = stats.zscore(heart_df)
abs_z_scores = np.abs(z_scores)
filtered_entries = (abs_z_scores < 7).all(axis=1)
new_heart_df = heart_df[filtered_entries]

The malign and benign counts of diagnosis after removing very far outliers are as follows

In [11]:
new_heart_df['diagnosis'].value_counts()

0    354
1    208
Name: diagnosis, dtype: int64

# EXPLORATORY ANALYSYS

In [ ]:

import seaborn as sns
sns.boxplot(new_heart_df['diagnosis'], new_heart_df['smoothness_mean'])

In [ ]:
plt=sns.distplot(new_heart_df['smoothness_mean'])


In [ ]:
new_heart_df['smoothness_mean'].hist(by=new_heart_df['diagnosis'])

In [ ]:
Behaviour = new_heart_df['diagnosis'].value_counts(normalize=False).plot(kind = 'barh')

In [ ]:
sns.scatterplot(new_heart_df['radius_mean'], new_heart_df['texture_mean'], hue = new_heart_df['diagnosis'])

In [ ]:
sns.boxplot(new_heart_df['diagnosis'], new_heart_df['perimeter_mean'])


In [ ]:
cond_plot = sns.FacetGrid(data=new_heart_df, hue='diagnosis')
cond_plot.map(sns.scatterplot, 'fractal_dimension_mean', 'perimeter_mean');

In [ ]:
cond_plot = sns.FacetGrid(data=new_heart_df, hue='diagnosis')
cond_plot.map(sns.scatterplot, 'concave_points_sd_error', 'radius_mean');

In [ ]:
sns.heatmap(new_heart_df.corr(),cmap="GnBu");

# Modelling

Before we do modeling on our dataset, we break the dataset into a dataset named 'data' which contains all the descriptive features and anther dataset named 'target' containing all the target values.

In [12]:
data = new_heart_df.drop(columns='diagnosis')
target = new_heart_df['diagnosis']

The shape of dataframe with descriptive variables as 'data' and the diagnosis as 'target' are as below

In [13]:
print("Shape of data")
print(data.shape)
print("Shape of target")
print(target.shape)

Shape of data
(562, 30)
Shape of target
(562,)


We will now scale all the desriptive variables using the min-max scaler for making it easier to model our data.

In [14]:
from sklearn import preprocessing
data_df=data.copy()
Data_scaler = preprocessing.MinMaxScaler()
Data_scaler.fit(data)
data = Data_scaler.fit_transform(data)
print("5 random rows of Dataframe after scaling")
pd.DataFrame(data, columns=data_df.columns).sample(5, random_state=999)

5 random rows of Dataframe after scaling


0    radius_mean  radius_sd_error  radius_worst  texture_mean  \
153     0.528633         0.373013      0.532210      0.389366   
549     0.157073         0.334122      0.151399      0.079848   
324     0.447107         0.196145      0.448341      0.308426   
401     0.289985         0.394319      0.280818      0.176122   
172     0.083502         0.160298      0.076856      0.039734   

0    texture_sd_error  texture_worst  perimeter_mean  perimeter_sd_error  \
153          0.531462       0.503801        0.434630            0.523857   
549          0.430351       0.190603        0.011774            0.055467   
324          0.381692       0.403123        0.282099            0.349950   
401          0.206554       0.218410        0.143533            0.092793   
172          0.349824       0.080063        0.000000            0.000000   

0    perimeter_worst  area_mean  ...  concavity_worst  concave_points_mean  \
153         0.460101   0.246420  ...         0.497619             0.348881   
549         0.369192   0.281171  ...         0.107937             0.289446   
324         0.364646   0.206403  ...         0.429762             0.265458   
401         0.262626   0.235468  ...         0.256746             0.399520   
172         0.334343   0.363943  ...         0.052857             0.133795   

0    concave_points_sd_error  concave_points_worst  symmetry_mean  \
153                 0.461121              0.343969       0.466420   
549                 0.097658              0.049926       0.365383   
324                 0.413047              0.272514       0.477646   
401                 0.230421              0.141863       0.150895   
172                 0.044441              0.022792       0.297365   

0    symmetry_sd_error  symmetry_worst  fractal_dimension_mean  \
153           0.312901        0.306239                0.520619   
549           0.089948        0.008590                0.076701   
324           0.407981        0.423590                0.680756   
401           0.161355        0.157094                0.192474   
172           0.041990        0.000000                0.000000   

0    fractal_dimension_sd_error  fractal_dimension_worst  
153                    0.177016                 0.146530  
549                    0.137394                 0.081202  
324                    0.286615                 0.237439  
401                    0.181944                 0.173619  
172                    0.202444                 0.153745  

[5 rows x 30 columns]

## Feature Selection and Ranking

We will now look at the best feature selection method which can be used on our data. We will use the roc_auc value score by fitting  on the K nearest neighbour estimator using different feature selecting technique. We cross validate on RepeatedStratifiedKFold with number of splits=5 and a repeatition of 3 times.The feature selection we will be using are \
Full set of features \
F-score method \
Mutual information method \

We will then check the performance comparison of these models by paired t-test and select the best feature selection method.This will then be used in pipeline for our modelling purpose.

In [15]:
#full set of features
from sklearn.model_selection import cross_val_score, RepeatedStratifiedKFold
from sklearn.neighbors import KNeighborsClassifier
cv_method = RepeatedStratifiedKFold(n_splits=5, n_repeats=3,random_state=999)
cv_results_full = cross_val_score(estimator=KNeighborsClassifier(n_neighbors=20),
                                  X=data,
                                  y=target,
                                  cv=cv_method,
                                  scoring='roc_auc')
print("roc_auc for full set of features")
print(cv_results_full.mean().round(3))

roc_auc for full set of features
0.99


In [16]:
# f-score method
from sklearn import feature_selection as fs
fs_fit_fscore = fs.SelectKBest(fs.f_classif, k=10)
fs_fit_fscore.fit_transform(data, target)
fs_indices_fscore = np.argsort(np.nan_to_num(fs_fit_fscore.scores_))[::-1][0:10]
best_features_fscore = data_df.columns[fs_indices_fscore].values
feature_importances_fscore = fs_fit_fscore.scores_[fs_indices_fscore]
cv_results_fscore = cross_val_score(estimator=KNeighborsClassifier(n_neighbors=20),
                                    X=data[:, fs_indices_fscore],
                                    y=target,
                                    cv=cv_method,
                                    scoring='roc_auc')
print("roc_auc for f-score method")
print(cv_results_fscore.mean().round(3))

roc_auc for f-score method
0.981


In [17]:
#mutual information method
fs_fit_mutual_info = fs.SelectKBest(fs.mutual_info_classif, k=10)
fs_fit_mutual_info.fit_transform(data, target)
fs_indices_mutual_info = np.argsort(fs_fit_mutual_info.scores_)[::-1][0:10]
best_features_mutual_info = data_df.columns[fs_indices_mutual_info].values
feature_importances_mutual_info = fs_fit_mutual_info.scores_[fs_indices_mutual_info]
cv_results_mutual_info = cross_val_score(estimator=KNeighborsClassifier(n_neighbors=20),
                                         X=data[:, fs_indices_mutual_info],
                                         y=target,
                                         cv=cv_method,
                                         scoring='roc_auc')
print("roc_auc value for mutual information method")
print(cv_results_mutual_info.mean().round(3))

roc_auc value for mutual information method
0.984


The roc_auc values for the 3 different feature selection method we have used with the same cross validation method,scoring method,and estimator is as given below.The higher value is that when using the full set of features.We use paired t-test to compare compare it to the other feature selection method results, to check if it is statistically signifant than the other 2.

In [18]:
print('Full Set of Features:', cv_results_full.mean().round(3))
print('F-Score:', cv_results_fscore.mean().round(3))
print('Mutual Information:', cv_results_mutual_info.mean().round(3))
print("paired t-test comparison values of mutual information method with others")
print(stats.ttest_rel(cv_results_full, cv_results_mutual_info).pvalue.round(3))
print(stats.ttest_rel(cv_results_full, cv_results_fscore).pvalue.round(3))

Full Set of Features: 0.99
F-Score: 0.981
Mutual Information: 0.984
paired t-test comparison values of mutual information method with others
0.0
0.0


We can see that the paired t-test returns a value which is less that 0.05 and thus we can say that using full feature set is statistically significant than choosing the other two methods. Thus for our modelling purpose we will go with the full descriptive feature set

## Train Test Splitting

We will now split our feature dataset and the target dataset into train and test partitions with a 7:3 ratio. We will be training the model on our train split datasets and testing it on the testing dataset.

In [19]:
from sklearn.model_selection import train_test_split
Data_train, Data_test, \
target_train, target_test = train_test_split(data, target, 
                                                    test_size = 0.3, random_state=999,
                                                    stratify = target)
print("Shape of training feature dataset")
print(Data_train.shape)
print("Shape of testing feature dataset")
print(Data_test.shape)
print("Shape of training target dataset")
print(target_train.shape)
print("Shape of testing feature dataset")
print(target_test.shape)

Shape of training feature dataset
(393, 30)
Shape of testing feature dataset
(169, 30)
Shape of training target dataset
(393,)
Shape of testing feature dataset
(169,)


For our modelling purpose we will use a 5 fold stratified cross validation method with a repetition of 3 times to train the model and hypertune the various model parameters. This will reduce the issues of overfitting of our models.

In [20]:
cv_method = RepeatedStratifiedKFold(n_splits=5, 
                                    n_repeats=3, 
                                    random_state=999)

The distribution of the target variable (diagnosis) is as below and can be observed that it is unevenly distributed. In such a case it is better use aus_roc scoring method for performance comparison of our model. Thus we will go forward with a performance scoring of auc_roc

In [21]:
target_train.value_counts(normalize=True)

0    0.631043
1    0.368957
Name: diagnosis, dtype: float64

## Modelling and Tuning Parameters of Different Model

### K-Nearest Neighbors(KNN) Model 

In [22]:
from sklearn.pipeline import Pipeline
from sklearn.feature_selection import SelectKBest
from sklearn.model_selection import GridSearchCV
pipe_KNN = Pipeline([('fselector', SelectKBest(fs.mutual_info_classif)), 
                     ('knn', KNeighborsClassifier())])

params_pipe_KNN = {'fselector__k': [data.shape[1]],
                   'knn__n_neighbors': [1, 5, 10, 15,30,50,70],
                   'knn__p': [1, 2,5]}

gs_pipe_KNN = GridSearchCV(estimator=pipe_KNN, 
                           param_grid=params_pipe_KNN, 
                           cv=cv_method,
                           scoring='roc_auc',
                           verbose=1) 

gs_pipe_KNN.fit(Data_train, target_train);

Fitting 15 folds for each of 21 candidates, totalling 315 fits


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 315 out of 315 | elapsed:   33.4s finished


In [23]:
print("Best set of parameters using KNN model")
print(gs_pipe_KNN.best_params_)
print("roc_auc value fitting with best parameters")
print(gs_pipe_KNN.best_score_)

Best set of parameters using KNN model
{'fselector__k': 30, 'knn__n_neighbors': 30, 'knn__p': 2}
roc_auc value fitting with best parameters
0.9936575181796855


### Gaussian Naive Bayes(NB) Model

In [24]:
from sklearn.naive_bayes import GaussianNB
from sklearn.preprocessing import PowerTransformer
Data_train_transformed = PowerTransformer().fit_transform(Data_train)

pipe_NB = Pipeline([('fselector', SelectKBest(fs.mutual_info_classif)), 
                     ('nb', GaussianNB())])

params_pipe_NB = {'fselector__k': [data.shape[1]],
                  'nb__var_smoothing': np.logspace(1,-3, num=20)}

gs_pipe_NB = GridSearchCV(estimator=pipe_NB, 
                          param_grid=params_pipe_NB, 
                          cv=cv_method,
                          refit=True,
                          n_jobs=-2,
                          scoring='roc_auc',
                          verbose=1) 

gs_pipe_NB.fit(Data_train_transformed, target_train);

Fitting 15 folds for each of 20 candidates, totalling 300 fits


[Parallel(n_jobs=-2)]: Using backend LokyBackend with 3 concurrent workers.
[Parallel(n_jobs=-2)]: Done  44 tasks      | elapsed:    4.2s
[Parallel(n_jobs=-2)]: Done 194 tasks      | elapsed:   12.2s
[Parallel(n_jobs=-2)]: Done 300 out of 300 | elapsed:   17.7s finished


In [25]:
print("Best set of parameters using NB model")
print(gs_pipe_NB.best_params_)
print("roc_auc value fitting with best parameters")
print(gs_pipe_NB.best_score_)

Best set of parameters using NB model
{'fselector__k': 30, 'nb__var_smoothing': 0.011288378916846895}
roc_auc value fitting with best parameters
0.9882120572366877


### Decision Tree(DT) Model

In [26]:
from sklearn.tree import DecisionTreeClassifier
pipe_DT = Pipeline([('fselector', SelectKBest(fs.mutual_info_classif)),
                    ('dt', DecisionTreeClassifier())])

params_pipe_DT = {'fselector__k': [data.shape[1]],
                  'dt__criterion': ['gini', 'entropy'],
                  'dt__max_depth': [2,3, 5, 7, 10],
                  'dt__min_samples_split': [2, 5, 15, 20, 30]}

gs_pipe_DT = GridSearchCV(estimator=pipe_DT, 
                          param_grid=params_pipe_DT, 
                          cv=cv_method,
                          refit=True,
                          n_jobs=-2,
                          scoring='roc_auc',
                          verbose=1) 
gs_pipe_DT.fit(Data_train, target_train);

Fitting 15 folds for each of 50 candidates, totalling 750 fits


[Parallel(n_jobs=-2)]: Using backend LokyBackend with 3 concurrent workers.
[Parallel(n_jobs=-2)]: Done  82 tasks      | elapsed:    4.3s
[Parallel(n_jobs=-2)]: Done 382 tasks      | elapsed:   20.6s
[Parallel(n_jobs=-2)]: Done 750 out of 750 | elapsed:   41.9s finished


In [27]:
print("Best set of parameters using DT model")
print(gs_pipe_DT.best_params_)
print("roc_auc value fitting with best parameters")
print(gs_pipe_DT.best_score_)

Best set of parameters using DT model
{'dt__criterion': 'gini', 'dt__max_depth': 5, 'dt__min_samples_split': 30, 'fselector__k': 30}
roc_auc value fitting with best parameters
0.9545615763546798


### Support-Vector Machines(SVM) MODEL

In [28]:
from sklearn.svm import SVC
pipe_SVM = Pipeline([('fselector', SelectKBest(fs.mutual_info_classif)),
                    ('svm', SVC())])

params_pipe_SVM = {'fselector__k': [data.shape[1]],
                   'svm__C': [ 1.e-02, 1.e-01, 1.e+00, 1.e+01, 1.e+02,],
                   'svm__gamma': [1.e-03, 1.e-02, 1.e-01, 1.e+00],
                   'svm__kernel': ['rbf','linear','poly']
                  }

gs_pipe_SVM = GridSearchCV(estimator=pipe_SVM, 
                          param_grid=params_pipe_SVM, 
                          cv=cv_method,
                          refit=True,
                          n_jobs=-2,
                          scoring='roc_auc',
                          verbose=1) 

gs_pipe_SVM.fit(Data_train, target_train);

Fitting 15 folds for each of 60 candidates, totalling 900 fits


[Parallel(n_jobs=-2)]: Using backend LokyBackend with 3 concurrent workers.
[Parallel(n_jobs=-2)]: Done  82 tasks      | elapsed:    4.0s
[Parallel(n_jobs=-2)]: Done 382 tasks      | elapsed:   21.0s
[Parallel(n_jobs=-2)]: Done 882 tasks      | elapsed:   51.9s
[Parallel(n_jobs=-2)]: Done 900 out of 900 | elapsed:   52.8s finished


In [29]:
print("Best set of parameters using SVM model")
print(gs_pipe_SVM.best_params_)
print("roc_auc value fitting with best parameters")
print(gs_pipe_SVM.best_score_)

Best set of parameters using SVM model
{'fselector__k': 30, 'svm__C': 10.0, 'svm__gamma': 0.1, 'svm__kernel': 'rbf'}
roc_auc value fitting with best parameters
0.9969786535303778


## Evaluating Model Performance

We will now fit the models with the best set of of parameters as obtained while modelling to the test data.We will then calculate the classification metrics, confusion matrix and the cross validated AUC score for the 4 models. 

In [34]:
from sklearn.model_selection import StratifiedKFold
import sklearn.metrics as metrics
cv_method_ttest = StratifiedKFold(n_splits=3, random_state=999)

cv_results_KNN = cross_val_score(estimator=gs_pipe_KNN.best_estimator_,
                                 X=Data_test,
                                 y=target_test, 
                                 cv=cv_method_ttest, 
                                 n_jobs=-2,
                                 scoring='roc_auc')
print("cross validated KNN AUC score")
print(cv_results_KNN.mean())
pred_KNN = gs_pipe_KNN.predict(Data_test)
print("\nClassification report for K-Nearest Neighbor") 
print(metrics.classification_report(target_test, pred_KNN))
print("\nConfusion matrix for K-Nearest Neighbor") 
print(metrics.confusion_matrix(target_test, pred_KNN))

cross validated KNN AUC score
0.9867598891408416

Classification report for K-Nearest Neighbor
              precision    recall  f1-score   support

           0       0.95      1.00      0.97       106
           1       1.00      0.90      0.95        63

    accuracy                           0.96       169
   macro avg       0.97      0.95      0.96       169
weighted avg       0.97      0.96      0.96       169


Confusion matrix for K-Nearest Neighbor
[[106   0]
 [  6  57]]


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:296: FutureWarning: Setting a random_state has no effect since shuffle is False. This will raise an error in 0.24. You should leave random_state to its default (None), or set shuffle=True.
  FutureWarning


In [36]:
Data_test_transformed = PowerTransformer().fit_transform(Data_test)

cv_results_NB = cross_val_score(estimator=gs_pipe_NB.best_estimator_,
                                X=Data_test_transformed,
                                y=target_test, 
                                cv=cv_method_ttest, 
                                n_jobs=-2,
                                scoring='roc_auc')
print("cross validated NB AUC score")
print(cv_results_NB.mean())
Data_test_transformed = PowerTransformer().fit_transform(Data_test)
pred_NB = gs_pipe_NB.predict(Data_test_transformed)
print("\n\nClassification report for Naive Bayes") 
print(metrics.classification_report(target_test, pred_NB))
print("\nConfusion matrix for Naive Bayes") 
print(metrics.confusion_matrix(target_test, pred_NB))

cross validated NB AUC score
0.987893675988914


Classification report for Naive Bayes
              precision    recall  f1-score   support

           0       0.97      0.95      0.96       106
           1       0.92      0.95      0.94        63

    accuracy                           0.95       169
   macro avg       0.95      0.95      0.95       169
weighted avg       0.95      0.95      0.95       169


Confusion matrix for Naive Bayes
[[101   5]
 [  3  60]]


In [32]:
cv_results_DT = cross_val_score(estimator=gs_pipe_DT.best_estimator_,
                                X=Data_test,
                                y=target_test, 
                                cv=cv_method_ttest, 
                                n_jobs=-2,
                                scoring='roc_auc')
print("cross validated DT AUC score")
print(cv_results_DT.mean())
pred_DT = gs_pipe_DT.predict(Data_test)
print("\n\nClassification report for Decision Tree") 
print(metrics.classification_report(target_test, pred_DT))
print("\nConfusion matrix for Decision Tree") 
print(metrics.confusion_matrix(target_test, pred_DT))

cross validated DT AUC score
0.9066956412194508


Classification report for Decision Tree
              precision    recall  f1-score   support

           0       0.90      0.92      0.91       106
           1       0.85      0.83      0.84        63

    accuracy                           0.88       169
   macro avg       0.88      0.87      0.87       169
weighted avg       0.88      0.88      0.88       169


Confusion matrix for Decision Tree
[[97  9]
 [11 52]]


In [33]:
cv_results_SVM = cross_val_score(estimator=gs_pipe_SVM.best_estimator_,
                                X=Data_test,
                                y=target_test, 
                                cv=cv_method_ttest, 
                                n_jobs=-2,
                                scoring='roc_auc')
print("cross validated SVM AUC score")
print(cv_results_SVM.mean())
pred_SVM = gs_pipe_SVM.predict(Data_test)
print("\n\nClassification report for SVM") 
print(metrics.classification_report(target_test, pred_SVM))
print("\nConfusion matrix for Decision Tree") 
print(metrics.confusion_matrix(target_test, pred_SVM))


cross validated SVM AUC score
0.9918493323255229


Classification report for SVM
              precision    recall  f1-score   support

           0       0.97      0.98      0.98       106
           1       0.97      0.95      0.96        63

    accuracy                           0.97       169
   macro avg       0.97      0.97      0.97       169
weighted avg       0.97      0.97      0.97       169


Confusion matrix for Decision Tree
[[104   2]
 [  3  60]]


From observing the AUC score we recieved after fitting the model to the test data we an see that the SVM model give us the highest and thus the best AUC score for the data. We can also observe from the performance metrics of the SVM model that it has a precision of 0.97 for predicting both the malignant and benign tumor diagnosis. The model also has a higher recall rate of 0.95 for the malignant tumor diagnosis which means that  We can also see that the AUC score is almost similar when fitted both on the training data and the testing data. Thus we can conclude that there is not much overfitting in our model. 